### PR TITLE
pkg/endpoint: Simplify search for C header file during restore

### PR DESCRIPTION
**Pull request by pchaigno**
Since we know the C header filename, we don't need to list the endpoint directory's content. We can simply check if the file exists.

*Extracted from the host firewall branch.*

@aanm Do you remember if there was a reason for listing the directory content instead of checking if the file exists? I feel like I'm missing something here. Commit [`2365922`](https://github.com/cilium/cilium/commit/2365922a02c83053fa08b4bd268861e71435e863#diff-bf5ec19662b7a3700c86fda28d3f305aR69) introduced this code.

### CODE REVIEW

**Comment by pchaigno** [comment]
test-me-please

**Comment by aanm** [comment]
> Since we know the C header filename, we don't need to list the endpoint directory's content. We can simply check if the file exists.
> 
> _Extracted from the host firewall branch._
> 
> @aanm Do you remember if there was a reason for listing the directory content instead of checking if the file exists? I feel like I'm missing something here. Commit [`2365922`](https://github.com/cilium/cilium/commit/2365922a02c83053fa08b4bd268861e71435e863#diff-bf5ec19662b7a3700c86fda28d3f305aR69) introduced this code.

@pchaigno I could never get to the bottom of it but it is as it sounds in the comment. The first "read" would return `os.FileNotFound`, but not the second attempt. That's why I kept it just to be safe, it doesn't hurt try reading the same file twice if the file didn't exist in the first place.

**Review by aanm — COMMENTED** [review]
testing

**Comment by pchaigno** [comment]
> > @aanm Do you remember if there was a reason for listing the directory content instead of checking if the file exists? I feel like I'm missing something here. Commit [`2365922`](https://github.com/cilium/cilium/commit/2365922a02c83053fa08b4bd268861e71435e863#diff-bf5ec19662b7a3700c86fda28d3f305aR69) introduced this code.
> 
> @pchaigno I could never get to the bottom of it but it is as it sounds in the comment. The first "read" would return `os.FileNotFound`, but not the second attempt. That's why I kept it just to be safe, it doesn't hurt try reading the same file twice if the file didn't exist in the first place.

I think we're talking about two different things. I was wondering if there was a special reason for listing the content instead of checking if the file exists. The second listing of the directory was introduced later. Checking directly if the file exists should hopefully remove the need for that second listing.

File: pkg/endpoint/restore.go
Comment (aanm) [inline pkg/endpoint/restore.go]: Can we attempt 3 times with a 100 ms sleep in between before failing? If this file is not found the endpoint is not restored.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/endpoint/restore.go
Comment (joestringer) [reply pkg/endpoint/restore.go]: ^^^ If we think this actually does something meaningful, it'd be nice to get evidence somehow. Could we add a metric (not necessarily promethus, but something we can fetch via `cilium metrics list`) that this actually occurred?

That way over time we can check this and see if we really need it or if the reason we had to put this in in the first place has gone away.

Comment (pchaigno) [reply pkg/endpoint/restore.go]: I agree that we need a way to get rid of this in the long term. It's either not an issue anymore or we need to reproduce and fix the underlying problem, but we don't want to carry this *ad aeternam* because we noticed it once.

Do we actually check metrics though? I was thinking we could print a debug message if the output changes with the 2nd/3rd calls and add that debug message to `badLogMessages`. That way we're more likely to notice this issue happening.

@aanm Do you remember in what context this was happening?

Comment (aanm) [reply pkg/endpoint/restore.go]: >I agree that we need a way to get rid of this in the long term. It's either not an issue anymore or we need to reproduce and fix the underlying problem, but we don't want to carry this ad aeternam because we noticed it once.

I agree but the problem is that if the file is not found the endpoint, and connectivity, of the pod will be broken.

>Do we actually check metrics though? I was thinking we could print a debug message if the output changes with the 2nd/3rd calls and add that debug message to badLogMessages. That way we're more likely to notice this issue happening.

We do in the sysdumps. We can also add a `Warning` message with the string "BUG".

>@aanm Do you remember in what context this was happening?

This happened to me when I was developing the restore code. The endpoints were not being restored and I added debug messages all over the place until I got the "File not found" which wouldn't fail on the 2nd attempt.

**Comment by pchaigno** [comment]
test-me-please

File: pkg/endpoint/restore.go
Comment (pchaigno) [reply pkg/endpoint/restore.go]: I pushed an update with 2 retries spaced by 100ms. I also added the BUG warning and a `badLogMessages` entry. I've run the `K8sUpdates` test 34 times locally to ensure the bug isn't actually too frequent, so that we won't block all PRs just after merging this.

File: pkg/endpoint/restore.go
Comment (aanm) [inline pkg/endpoint/restore.go]: Do we need to do the Stat if the file was found? I had the feeling that this only matters to be retried if the file is not found.

**Review by christarazi — APPROVED** [review]
LGTM, assuming @aanm's comment is addressed

**Comment by pchaigno** [comment]
test-me-please

**Pull request merged**
